track_analyzing/track_analyzer/cmd_table.cpp

```diff
@@ -0,0 +1,273 @@
+      else
+        out << " ";
+
+      out << it.first.ToString() << ": " << speedInfo.GetDistance() << " / " << speedInfo.GetTime();
+    }
+
+    return out.str();
+  }
+
+private:
+  map<MoveType, MoveInfo> m_moveInfos;
+};
+
+class MatchedTrackPointToMoveType final
+{
+public:
+  MatchedTrackPointToMoveType(string const & mwmFile)
+    : m_featuresVector(FilesContainerR(make_unique<FileReader>(mwmFile)))
+  {
+  }
+
+  MoveType GetMoveType(MatchedTrackPoint const & point)
+  {
+    return MoveType(GetRoadType(point.GetSegment().GetFeatureId()),
+                    static_cast<traffic::SpeedGroup>(point.GetDataPoint().m_traffic));
+  }
+
+private:
+  uint32_t GetRoadType(uint32_t featureId)
+  {
+    if (featureId == m_prevFeatureId)
+      return m_prevRoadType;
+
+    FeatureType feature;
+    m_featuresVector.GetVector().GetByIndex(featureId, feature);
+    feature.ParseTypes();
+
+    m_prevFeatureId = featureId;
+    m_prevRoadType = m_carModelTypes.GetType(feature);
+    return m_prevRoadType;
+  }
+
+  CarModelTypes const m_carModelTypes;
+  FeaturesVectorTest m_featuresVector;
+  uint32_t m_prevFeatureId = numeric_limits<uint32_t>::max();
+  uint32_t m_prevRoadType = numeric_limits<uint32_t>::max();
+};
+}  // namespace
+
+namespace track_analyzing
+{
+void CmdTagsTable(string const & filepath, string const & trackExtension, string const & mwmFilter,
+                  string const & userFilter)
+{
+  cout << "mwm user track_idx start length time speed ... type: meters / seconds" << endl;
+
+  storage::Storage storage;
+  auto numMwmIds = CreateNumMwmIds(storage);
+
+  auto processMwm = [&](string const & mwmName, UserToMatchedTracks const & userToMatchedTracks) {
+    if (IsFiltered(mwmFilter, mwmName))
+      return;
+
+    shared_ptr<IVehicleModel> vehicleModel = CarModelFactory().GetVehicleModelForCountry(mwmName);
+    string const mwmFile =
+        my::JoinPath(GetPlatform().WritableDir(), to_string(storage.GetCurrentDataVersion()),
+                     mwmName + DATA_FILE_EXTENSION);
+    MatchedTrackPointToMoveType pointToMoveType(mwmFile);
+    Geometry geometry(GeometryLoader::CreateFromFile(mwmFile, vehicleModel));
+
+    for (auto uIt : userToMatchedTracks)
+    {
+      string const & user = uIt.first;
+      if (IsFiltered(userFilter, user))
+        continue;
+
+      for (size_t trackIdx = 0; trackIdx < uIt.second.size(); ++trackIdx)
+      {
+        MatchedTrack const & track = uIt.second[trackIdx];
+        uint64_t const start = track.front().GetDataPoint().m_timestamp;
+        uint64_t const timeElapsed = track.back().GetDataPoint().m_timestamp - start;
+        double const length = CalcTrackLength(track, geometry);
+        double const speed = CalcSpeedKMpH(length, timeElapsed);
+
+        MoveTypeAggregator aggregator;
+
+        MoveType currentType(numeric_limits<uint32_t>::max(), traffic::SpeedGroup::Count);
+        size_t subTrackBegin = 0;
+
+        for (size_t i = 0; i < track.size(); ++i)
```

Comment:
```c++
size_t i = 0;
while (i < track.size()) {
  auto const t = pointToMoveType.GetMoveType(track[i]);

  size_t j = i;
  while (j < track.size() && pointToMoveType.GetMoveType(track[j]) == t)
    ++j;
  aggregator.Add(currentType, track, i, j, geometry);
  i = j;
}
```
This form of two-pointers method is well-known, quite convenient and more readable.